### PR TITLE
Feature/external event building

### PR DESCRIPTION
**Pull request by a-d-smith**
Required changes to go along with LArPandora feature

### CODE REVIEW

File: larpandoracontent/LArControlFlow/MasterAlgorithm.cc [hunk 1]
```diff
@@ -210,13 +211,13 @@ StatusCode MasterAlgorithm::InitializeWorkerInstances()
             m_crWorkerInstances.push_back(this->CreateWorkerInstance(*(mapEntry.second), gapList, m_crSettingsFile));
```
Comment (PandoraPFA) [inline larpandoracontent/LArControlFlow/MasterAlgorithm.cc]: Might make sense to call these instances CRWorkerInstance1, etc.

File: larpandoracontent/LArControlFlow/MasterAlgorithm.cc [hunk 2]
```diff
@@ -210,13 +211,13 @@ StatusCode MasterAlgorithm::InitializeWorkerInstances()
...
 
         if (m_shouldRunSlicing)
-            m_pSlicingWorkerInstance = this->CreateWorkerInstance(larTPCMap, gapList, m_slicingSettingsFile);
+            m_pSlicingWorkerInstance = this->CreateWorkerInstance(larTPCMap, gapList, m_slicingSettingsFile, "SliceWorker");
 
         if (m_shouldRunNeutrinoRecoOption)
-            m_pSliceNuWorkerInstance = this->CreateWorkerInstance(larTPCMap, gapList, m_nuSettingsFile);
+            m_pSliceNuWorkerInstance = this->CreateWorkerInstance(larTPCMap, gapList, m_nuSettingsFile, "NeutrinoWorker");
```
Comment (PandoraPFA) [inline larpandoracontent/LArControlFlow/MasterAlgorithm.cc]: SliceNuWorker?

File: larpandoracontent/LArControlFlow/MasterAlgorithm.cc [hunk 3]
```diff
@@ -210,13 +211,13 @@ StatusCode MasterAlgorithm::InitializeWorkerInstances()
...
 
         if (m_shouldRunCosmicRecoOption)
-            m_pSliceCRWorkerInstance = this->CreateWorkerInstance(larTPCMap, gapList, m_crSettingsFile);
+            m_pSliceCRWorkerInstance = this->CreateWorkerInstance(larTPCMap, gapList, m_crSettingsFile, "CosmicRayWorker");
```
Comment (PandoraPFA) [inline larpandoracontent/LArControlFlow/MasterAlgorithm.cc]: SliceCRWorker?

File: larpandoracontent/LArControlFlow/MasterAlgorithm.h [hunk 4]
```diff
@@ -262,23 +262,25 @@ class MasterAlgorithm : public pandora::ExternallyConfiguredAlgorithm
      *  @param  larTPC the lar tpc
      *  @param  gapList the gap list
      *  @param  settingsFile the pandora settings file
+     *  @param  name the pandora instance name
      *
      *  @return the address of the pandora instance
      */
     const pandora::Pandora *CreateWorkerInstance(const pandora::LArTPC &larTPC, const pandora::DetectorGapList &gapList,
-        const std::string &settingsFile) const;
+        const std::string &settingsFile, const std::string &name = "") const;
```
Comment (PandoraPFA) [inline larpandoracontent/LArControlFlow/MasterAlgorithm.h]: Can remove implicit empty string argument if all worker instances are given a name.

File: larpandoracontent/LArControlFlow/MasterAlgorithm.h [hunk 5]
```diff
@@ -262,23 +262,25 @@ class MasterAlgorithm : public pandora::ExternallyConfiguredAlgorithm
...
      *  @param  larTPCMap the lar tpc map
      *  @param  gapList the gap list
      *  @param  settingsFile the pandora settings file
+     *  @param  name the pandora instance name
      *
      *  @return the address of the pandora instance
      */
     const pandora::Pandora *CreateWorkerInstance(const pandora::LArTPCMap &larTPCMap, const pandora::DetectorGapList &gapList,
-        const std::string &settingsFile) const;
+        const std::string &settingsFile, const std::string &name = "") const;
```
Comment (PandoraPFA) [inline larpandoracontent/LArControlFlow/MasterAlgorithm.h]: And again.

File: larpandoracontent/LArControlFlow/MasterAlgorithm.cc [hunk 6]
```diff
@@ -210,13 +211,13 @@ StatusCode MasterAlgorithm::InitializeWorkerInstances()
             m_crWorkerInstances.push_back(this->CreateWorkerInstance(*(mapEntry.second), gapList, m_crSettingsFile));
 
         if (m_shouldRunSlicing)
-            m_pSlicingWorkerInstance = this->CreateWorkerInstance(larTPCMap, gapList, m_slicingSettingsFile);
+            m_pSlicingWorkerInstance = this->CreateWorkerInstance(larTPCMap, gapList, m_slicingSettingsFile, "SliceWorker");
```
Comment (PandoraPFA) [inline larpandoracontent/LArControlFlow/MasterAlgorithm.cc]: SlicingWorker?

File: larpandoracontent/LArControlFlow/MasterAlgorithm.cc [hunk 7]
```diff
@@ -380,8 +381,14 @@ StatusCode MasterAlgorithm::TagCosmicRayPfos(const PfoToFloatMap &stitchedPfosTo
...
     for (const Pfo *const pPfo : nonStitchedParentCosmicRayPfos)
     {
-        if (ambiguousPfos.end() == std::find(ambiguousPfos.begin(), ambiguousPfos.end(), pPfo))
+        const bool isClearCosmic(ambiguousPfos.end() == std::find(ambiguousPfos.begin(), ambiguousPfos.end(), pPfo));
+
+        if (isClearCosmic)
             clearCosmicRayPfos.push_back(pPfo);
+    
+        PandoraContentApi::ParticleFlowObject::Metadata metadata;
+        metadata.m_propertiesToAdd["IsClearCosmic"] = isClearCosmic;
```
Comment (PandoraPFA) [inline larpandoracontent/LArControlFlow/MasterAlgorithm.cc]: Implicit bool to float conversion here. Might want to make that explicit, or use conditional operator ( ? : )

File: larpandoracontent/LArControlFlow/MasterAlgorithm.cc [hunk 8]
```diff
@@ -501,6 +508,13 @@ StatusCode MasterAlgorithm::RunSliceReconstruction(SliceVector &sliceVector, Sli
             PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraApi::ProcessEvent(*m_pSliceNuWorkerInstance));
             PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraApi::GetCurrentPfoList(*m_pSliceNuWorkerInstance, pSliceNuPfos));
             nuSliceHypotheses.push_back(*pSliceNuPfos);
+    
+            for (const ParticleFlowObject *const pPfo : *pSliceNuPfos)
+            {
+                PandoraContentApi::ParticleFlowObject::Metadata metadata;
+                metadata.m_propertiesToAdd["SliceIndex"] = sliceCounter;
```
Comment (PandoraPFA) [inline larpandoracontent/LArControlFlow/MasterAlgorithm.cc]: So SliceIndices start at 1? That's OK, if consistent throughout all treatment. Helps to keep code clean here.

**Comment by a-d-smith** [comment]
Thanks for your comments here! I will make these changes

File: larpandoracontent/LArControlFlow/MasterAlgorithm.cc (hunk 1 above)
Comment (PandoraPFA) [reply larpandoracontent/LArControlFlow/MasterAlgorithm.cc]: You could here use the volume id of the associated LArTPC instance to make sure that you give each instance a meaningful name. (Obviously be aware that multiple Pandora instances are allowed to have the same name, and there is no mechanism to check names for all instances to ensure uniqueness, so it's your responsibility to give these instances useful names!)

File: larpandoracontent/LArControlFlow/MasterAlgorithm.cc (hunk 8 above)
Comment (a-d-smith) [reply larpandoracontent/LArControlFlow/MasterAlgorithm.cc]: Yes! I index from 1 here to be consistent with couts in the master instance

**Pull request merged**